File: clients/rn/src/engine/between/BetweenHeadlineSyncClient.ts
/**
 * HeadlineSyncClient backed by the Between network (NATS).
 * Posts headlines to NATS and receives headlines from siblings.
 *
 * Subject: "between.household.{familyId}.{nodeId}.soul.headlines"
 *
 * Falls back to local-only mode if NATS is not connected.
 *
 * and.
 */

import type { Headline, HeadlineCallback } from '../soul/HeadlineSyncClient';
import type { BetweenClient } from './BetweenClient';

export class BetweenHeadlineSyncClient {
  private headlines = new Map<string, Headline>();
  private listeners: HeadlineCallback[] = [];
  private unsubscribe: (() => void) | null = null;

  constructor(
    private readonly between: BetweenClient,
    private readonly nodeId: string,
    private readonly familyId: string,
  ) {}

  /** Start listening for headlines from the Between network. */
  startListening(): void {
    const subject = this.headlineSubject('*');
    this.unsubscribe = this.between.subscribe(subject, (_subject, data) => {
      try {
        const headline = JSON.parse(new TextDecoder().decode(data)) as Headline;
        // Don't echo our own headlines back
        if (headline.budDid !== this.nodeId) {
          this.receiveHeadline(headline);
        }
      } catch {
        // Malformed headline — skip
      }
    });
  }

  /** Stop listening for headlines. */
  stopListening(): void {
    this.unsubscribe?.();
    this.unsubscribe = null;
  }

  /** Post a headline from this bud. */
  postHeadline(headline: Headline): void {
    this.headlines.set(headline.budDid, headline);

    // Publish to Between if connected
    if (this.between.isConnected) {
      try {
        const data = new TextEncoder().encode(JSON.stringify(headline));
        this.between.publish(this.headlineSubject(this.nodeId), data);
      } catch {
        // Publish failure is non-fatal
      }
    }

    // Notify local listeners
    for (const listener of this.listeners) {
      listener(headline);
    }
  }

  /** Get latest headlines sorted by timestamp (newest first). */
  latestHeadlines(): Headline[] {
    return [...this.headlines.values()].sort((a, b) => b.timestamp - a.timestamp);
  }

  /** Subscribe to incoming headlines. Returns unsubscribe function. */
  onHeadlineReceived(callback: HeadlineCallback): () => void {
    this.listeners.push(callback);
    return () => {
      this.listeners = this.listeners.filter(l => l !== callback);
    };
  }

  private receiveHeadline(headline: Headline): void {
    this.headlines.set(headline.budDid, headline);
    for (const listener of this.listeners) {
      listener(headline);
    }
  }

  private headlineSubject(src: string): string {
    return `between.household.${this.familyId}.${src}.soul.headlines`;
  }
}

File: clients/kmp/shared/src/commonMain/kotlin/org/wyrdsekai/app/engine/between/BetweenHeadlineSyncClient.kt
package org.wyrdsekai.app.engine.between

import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.wyrdsekai.app.engine.soul.Headline
import org.wyrdsekai.app.engine.soul.HeadlineSyncClient

/**
 * HeadlineSyncClient backed by the Between network.
 * Posts headlines to NATS and receives headlines from siblings.
 *
 * Subject: "between.household.{nodeId}.*.soul.headlines"
 *
 * Falls back to local-only mode if Between is not connected.
 */
class BetweenHeadlineSyncClient(
    private val between: BetweenClient,
    private val nodeId: String,
    private val familyId: String,
) : HeadlineSyncClient {

    private val headlines = mutableMapOf<String, Headline>()
    private val listeners = mutableListOf<(Headline) -> Unit>()
    private var unsubscribe: (() -> Unit)? = null

    private val json = Json { ignoreUnknownKeys = true }

    /** Start listening for headlines from the Between network. */
    fun startListening() {
        val subject = headlineSubject("*")
        unsubscribe = between.subscribe(subject) { _, data ->
            try {
                val headline = json.decodeFromString<Headline>(data.decodeToString())
                // Don't echo our own headlines back
                if (headline.budDid != nodeId) {
                    receiveHeadline(headline)
                }
            } catch (_: Exception) {
                // Malformed headline — skip
            }
        }
    }

    /** Stop listening for headlines. */
    fun stopListening() {
        unsubscribe?.invoke()
        unsubscribe = null
    }

    override suspend fun postHeadline(headline: Headline) {
        headlines[headline.budDid] = headline

        // Publish to Between if connected
        if (between.isConnected) {
            try {
                val data = json.encodeToString(headline).encodeToByteArray()
                between.publish(headlineSubject(nodeId), data)
            } catch (_: Exception) {
                // Publish failure is non-fatal — headline is still cached locally
            }
        }

        // Notify local listeners
        listeners.forEach { it(headline) }
    }

    override fun latestHeadlines(): List<Headline> =
        headlines.values.sortedByDescending { it.timestamp }

    override fun onHeadlineReceived(callback: (Headline) -> Unit) {
        listeners.add(callback)
    }

    private fun receiveHeadline(headline: Headline) {
        headlines[headline.budDid] = headline
        listeners.forEach { it(headline) }
    }

    private fun headlineSubject(src: String): String =
        "between.household.$familyId.$src.soul.headlines"
}
